### PR TITLE
1.4 Preview/Change Management System Part 1

### LINKED ISSUE #624: QA: Naming of directories on linux
tModLoader is using spaces ` ` on linux which is making it pita to handle -> Requesting name change alike `Mod Sources` -> `ModSources`

```
(base) kreyren@rogisek:~/.local/share/Terraria/ModLoader$ ls
 config.json     'input profiles.json'   Mods           Players        Worlds
 favorites.json   Logs                  'Mod Sources'   TexturePacks
```

### LINKED ISSUE #2073: Provide a smooth transition for breaking changes throughout ongoing development
# Goal
When we transition from out of 1.4 alpha and encourage modders and users to transition, API breaking changes will still come, but they need to come with a predictable preview, some way for modders to prepare for them, and a way to shield users who just want to keep their current playthrough safe.

# Proposal
### Github branches:
- `1.4-stable` - bugfixes only, updates once a month
- `1.4-preview` - breaking changes before the 20th of the month. Modders should develop here to use new features and prepare for the monthly update.
- `1.4-dev` - github development and testing. Will track preview before the 20th, and accumulate breaking changes after until the next month rollover.

### Versioning:
- Versioning will be `yyyy.mm.ff.pp` preview and stable will always have different `yyyy.mm` versions. The release cycle will be monthly, focused around transitioning this month's preview to stable.
- The tML `BuildInfo.BuildPurpose` will be set to dev/preview/release based on the branch. Only preview and stable will deploy to steam, and allow mod publishing.
- `1.4-dev` will not be versioned. Versions are identified by commit sha.
- Every commit to `1.4-preview` will increment the `.ff` version.
- Every commit to `1.4-stable` aside from the auto-merge action, will increment the `.pp` version.

### Mod publishing:
- A mod will publish to steam inside a subfolder matching the yyyy.mm version of the tML build.
- Only two different yyyy.mm folders can exist simultaneously.
- tML will load the newest version of the mod it can.

Effectively, because preview and stable will always have different yyyy.mm versions, this means that modders can publish and maintain two versions of their mods. Stable and preview. When tML stable version updates, players will automatically start loading the newer version of the mod which has been prepared, and modders can continue to develop next month's release against the new preview version.

### Github Actions and guidelines
- An action will be created, preview -> stable
    Prereqs: 
        - `1.4-preview` is tagged with `approved-for-stable`
        - `1.4-stable` is a parent of `1.4-preview` (allowing a fast-forward merge)
        - `1.4-preview` is tagged with a newer `yyyy.mm` version than `1.4-stable`
        - `1.4-preview` `yyyy.mm` version is < the current `yyyy.mm`
    
    Steps:
        - fast-forward `1.4-stable` to `1.4-preview`
        - set `1.4-preview` version to `yyyy.mm.0.0`
        - merge `1.4-dev` to `1.4-preview` (if this fails, the workflow won't abort)
        - run all steam deploy actions
        
    This action will be automatically run at midnight GMT on the 1st of each month. If it fails, a member of the tML team may re-run it, but due to the prereqs, each preview yyyy.mm version can only be merged to stable once.
    
- PRs will target `1.4-dev` by default. All PRs will be reviewed in some form before being accepted in to `1.4-dev`.
- Before midnight GMT on the 20th of the month, `1.4-preview` will automatically track `1.4-dev`, via fast-forward merges.
    After the 20th, commits to `1.4-preview` will be focused on non-breaking changes and stabilising the preview.
- As mentioned above, commits will increment the version numbers automatically.
- Bugfixes to `1.4-stable` will have to be merged back into `1.4-qa`/`1.4-preview` manually
   
### How modders find out about and react to breaking changes:
- Through a discord announcement channel, breaking change PRs must include porting instructions.
- Modders should test their mod on the preview after the 20th of each month, make any ports, and publish a new version if required. We hope the predictable release cycle and good communications from the tML team via discord can keep this a smooth process.
- **Bonus** Through a subscribe-able endpoint that can be used to trigger their own CI based on the results of our Action
  - How? Auto-rebuild and publish for each preview version? Needs more thought...

### PR DESCRIPTION
**Pull request by Solxanich**
tl;dr: a description for implementation of /closes #2073 

This PR introduces changes to the workflow on releases for tModLoader.
We will be formally moving over to a continuous release style of product, with a consistent release schedule.
Given the complex development of the API, we expect further changes to keep coming, however the bulk of major changes have been completed, so our approach moving forward will be to strike a balance in maintaining a released product as opposed to holding it back much longer.

**What Is Changing for Everyone**
Several folder and file locations have been changed to facilitate the new implementation.
First and foremost, the Logs folder has been moved to be in the Install Directory - you can now find all logs there.
Second, the ModLoader/Beta has been split in to three new folders corresponding to each 'stream' -> 
 - tModLoader (For General 1.4 tmodloader)
 - tModLoader-dev (for contributors using a local build of tml)
 - tModLoader-preview (the preview branch of tmodloader, see below)
Finally, all the folders in the above have been converted to remove spaces. For example, Mod Sources is now ModSources. This rename brings us in line with unix expectations.
All of these folder changes will occur automatically on first run after merge - for those who have symlinks in the Mod Sources folder - these will be carried over to the Preview set of files, but won't carryover elsewise and instead copies will be left there. Please be wary accordingly.

**What Is Changing as a Modder**
tl;dr: With the merge of the 1.4 Preview PR, we now fully recommend all modders begin developing on the 1.4 Preview branch in preparation for the move to a stable release. The first Stable release is being tentatively targeted for May.

The 1.4 Preview Branch contains the set of changes that will be deployed in the upcoming stable releases. It will take a copy of the 1,4 development branch and keep the changes on that up to the 21st of the month. 
After the 21st of the month, the preview version moves to be fully considered the candidate stable release, with no further planned changes. As such, this should give you a guaranteed period of 1 week to make changes to your mods ahead of stable release. Further, by developing against it, you should be caught up in general with the rare tML breaking changes.
Once Stable is released for the new month, Players will automatically received the updated version of your mod - you do NOT need to republish the mod. 

**What Is Changing as a Player?** 
tl;dr: With the merge of the 1.4 Preview PR, nothing has changed for players at this time. More to come later

Later in April, we will setup the 1.4 stable branch - this branch will be the branch that we recommend players use for a stable experience. It will usually update once a month, during the first week of the month. Sometime in May we hope to set this as the default branch, and move 1.3 to be its own Steam Beta option.
The 1.4 Preview branch/beta will be similar to the 'Alpha' experience - it will contain a copy of all the latest development changes up to a certain point monthly -> your favorite mod developers will be developing their mods against this branch so use at your own discretion for latest content.

**What Is Changing as a Contributor?** 
tl;dr: Continue to create PRs that target 1.4. Nothing has changed for contributors.

The 1.4-preview branch will be fast forwarded with any commits made against the 1.4 branch. As such, you do not need to ever really interact with the 1.4-preview branch outside of emergency bug fixes.
In the event of an emergency bug fix, the process will be determined at time of based on the circumstances; options may include cherry-picking commits or manual fast forwarding.

---------------------------------------------------------
Missing stuff:

- [x] Better Auto-versioning tracking
- [x] Completing code for loading from folders
- [x] Completing cleanup code for cleaning out old folders/root when two newer tml version are present
- [x] Ability to prevent Monthly Merge from running automatically
- [ ] Lots of Testing
- [ ] Steam Deployment changes
- [ ] New, Ready to use Github branches
- [ ] All the stuff under how Modders get notified
- [x] New save directory system
- [ ] Ability to Port save data between streams
- [x] Fixes #624
- [x] Centralized logging folder
- [x] ExampleMod works for CI publishing

### CODE REVIEW

File: patches/tModLoader/Terraria/ModLoader/ModLoader.cs [hunk 1]
```diff
@@ -36,7 +36,7 @@ public static class ModLoader
 		public static bool AlphaWelcomed;
 		public static bool ShowFirstLaunchWelcomeMessage;
 
-		public static string versionedName => (ModCompile.DeveloperMode || !BuildInfo.IsRelease) ? BuildInfo.versionedNameDevFriendly : BuildInfo.versionedName;
+		public static string versionedName => (ModCompile.DeveloperMode || !BuildInfo.IsRelease && !BuildInfo.IsBeta) ? BuildInfo.versionedNameDevFriendly : BuildInfo.versionedName;
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/ModLoader.cs]: We should change this to instead something like the version being `undefined` if it wasn't built from CI. Alternatively, if people want to checkout and build tags, we could use `git name-rev --name-only --tags HEAD` to return the tag of the current commit, which apparently returns `undefined` for non-tagged versions.

File: patches/tModLoader/Terraria/Program.cs.patch [hunk 2]
```diff
@@ -87,10 +87,14 @@
...
  			try {
  				Console.OutputEncoding = Encoding.UTF8;
-@@ -138,32 +_,68 @@
+@@ -138,32 +_,97 @@
  			}
  		}
  
++		public const string PreviewFolder = "tModLoader-Preview";
++		public const string ReleaseFolder = "tModLoader";
++		public const string DevFolder = "tModLoader-Dev";
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Program.cs.patch]: prefer lower case for `-dev` and `-preview`

File: patches/tModLoader/Terraria/Program.cs.patch [hunk 3]
```diff
@@ -100,16 +104,41 @@
...
++
++				FileUtilities.CopyFolder(newPath, Path.Combine(SavePath, DevFolder));
++				FileUtilities.CopyFolder(newPath, Path.Combine(SavePath, ReleaseFolder));
++			}
++
++			var fileFolder =
++				BuildInfo.IsRelease ? ReleaseFolder :
++				BuildInfo.IsBeta ? PreviewFolder :
++				DevFolder;
++			SavePath = Path.Combine(SavePath, fileFolder);
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Program.cs.patch]: this should definitely be its own function

File: patches/tModLoader/Terraria/Program.cs.patch [hunk 4]
```diff
@@ -100,16 +104,41 @@
  				args = Utils.ConvertMonoArgsToDotNet(args);
  
 +			LaunchParameters = Utils.ParseArguements(args);
++			
 +			SavePath =
 +				LaunchParameters.ContainsKey("-tmlsavedirectory") ? LaunchParameters["-tmlsavedirectory"] :
-+				LaunchParameters.ContainsKey("-savedirectory") ? Path.Combine(LaunchParameters["-savedirectory"], "ModLoader") :
-+				Platform.Get<IPathService>().GetStoragePath($"Terraria{Path.DirectorySeparatorChar}ModLoader{(!BuildInfo.IsRelease ? Path.DirectorySeparatorChar + "Beta" : "")}");
++				LaunchParameters.ContainsKey("-savedirectory") ? LaunchParameters["-savedirectory"] :
++				Platform.Get<IPathService>().GetStoragePath($"Terraria");
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Program.cs.patch]: `-savedirectory` is the vanilla command, and should mirror the directory structure like you have here. `-tmlsavedirectory` is meant to specify the exact tML save path (so no preview/dev/release logic should be applied). I think this has value, otherwise we should remove `-tmlsavedirectory`

File: patches/tModLoader/Terraria/Program.cs.patch (hunk 3 above)
Comment (Chicken-Bones) [reply patches/tModLoader/Terraria/Program.cs.patch]: You'll also need `try/catch` around all the file operations, in case of file in use errors. We should probably just display an error dialog if we can't move the dirs

Comment (Chicken-Bones) [reply patches/tModLoader/Terraria/Program.cs.patch]: If we centralize the logs folder, this can also be done after logging is initialized.

File: patches/tModLoader/Terraria/Program.cs.patch [hunk 5]
```diff
@@ -100,16 +104,41 @@
...
++			}
++
++			var fileFolder =
++				BuildInfo.IsRelease ? ReleaseFolder :
++				BuildInfo.IsBeta ? PreviewFolder :
++				DevFolder;
++			SavePath = Path.Combine(SavePath, fileFolder);
 +
 +			if (File.Exists("savehere.txt"))
 +				SavePath = "ModLoader"; // Fallback for unresolveable antivirus/onedrive issues. Also makes the game portable I guess.
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Program.cs.patch]: Maybe this should be `SavePath = release ? "tModLoader" : preview ? "tModLoader-preview" : "tModLoader-dev"`?

File: patches/tModLoader/Terraria/Social/Steam/WorkshopSocialModule.TML.cs [hunk 6]
```diff
@@ -89,10 +94,12 @@ public partial class WorkshopSocialModule
 
 			buildData["workshopdeps"] = workshopDeps;
 
-			string contentFolderPath = GetTemporaryFolderPath() + modFile.Name;
+			string contentFolderPath = GetTemporaryFolderPath() + modFile.Name + "/" + BuildInfo.tMLVersion.Major + "." + BuildInfo.tMLVersion.Minor;
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Social/Steam/WorkshopSocialModule.TML.cs]: string interpolation preferred

File: patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs [hunk 7]
```diff
@@ -319,5 +322,70 @@ internal static class ModOrganizer
...
+					}
+				}
+				else {
+					val = fileName;
+					break;
+				}
+			}
+
+			File.Delete(val);
+		}
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs]: is it possible to use `LINQ` to make these more readable?

File: patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs [hunk 8]
```diff
@@ -44,40 +45,42 @@ internal static class ModOrganizer
...
-					if (!modsDirCache.TryGetValue(fileName, out var mod) || mod.lastModified != lastModified) {
-						try {
-							var modFile = new TmodFile(fileName);
-
-							using (modFile.Open()) {
-								mod = new LocalMod(modFile) {
-									lastModified = lastModified
-								};
-							}
+				var fileName = GetActiveTmodInRepo(repo);
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs:46]: I think this breaks loading multiple (or any) mods from the `Mods` (dev builds) repo

File: patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs [hunk 9]
```diff
@@ -319,5 +322,70 @@ internal static class ModOrganizer
...
+		internal static string GetActiveTmodInRepo(string repo) {
+			Version tmodVersion = new Version(BuildInfo.tMLVersion.Major, BuildInfo.tMLVersion.Minor);
+			string[] tmods = Directory.GetFiles(repo, "*.tmod", SearchOption.AllDirectories);
+			if (tmods.Length == 1)
+				return tmods[0];
+
+			string val = null;
+			Version currVersion = null;
+			foreach (string fileName in tmods) {
+				var regex = new Regex(@"(\d\d\d\d[.]\d\d)");
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs]: it looks like you're putting the tMod version in the mod filename? This might be a good idea, I think my original expectation was that we'd put them in subfolders.

I'd prefer if we actually read the mod file header, but there isn't actually a `TmodFile` method to just read to the tML version. Though it's the first string after the magic `TMOD` bytes so it's not expensive to get.

File: patches/tModLoader/Terraria/Social/Steam/WorkshopSocialModule.cs.patch [hunk 10]
```diff
@@ -9,3 +9,12 @@
  	{
  		private WorkshopHelper.UGCBased.Downloader _downloader;
  		private WorkshopHelper.UGCBased.PublishedItemsFinder _publishedItems;
+@@ -24,7 +_,7 @@
+ 			_publisherInstances = new List<WorkshopHelper.UGCBased.APublisherInstance>();
+ 			base.ProgressReporter = new WorkshopProgressReporter(_publisherInstances);
+ 			base.SupportedTags = new SupportedWorkshopTags();
+-			_contentBaseFolder = Main.SavePath + Path.DirectorySeparatorChar + "Workshop";
++			_contentBaseFolder = Directory.GetParent(Main.SavePath).ToString() + Path.DirectorySeparatorChar + "Workshop";
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Social/Steam/WorkshopSocialModule.cs.patch]: prefer `Path.Combine` 
It looks like this is only used for `GetTemporaryFolderPath` so there's no reason for it to be common to `Terraria`, or across instances?

**Review by Chicken-Bones — CHANGES_REQUESTED** [review]

File: .github/workflows/MonthlyMerge.yml [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    steps:
+        - name: Check out base branch
+          if: ${{ steps.date.outputs.date }}<20
+          uses: actions/checkout@v2
+          with:
+            persist-credentials: false
+            fetch-depth: 10
+              
+        - name: Merge
+          if: ${{ steps.date.outputs.date }}<20
```
Comment (Chicken-Bones) [inline .github/workflows/MonthlyMerge.yml]: leftover condition?

File: patches/tModLoader/Terraria/Social/Steam/WorkshopSocialModule.cs.patch (hunk 10 above)
Comment (Solxanich) [reply patches/tModLoader/Terraria/Social/Steam/WorkshopSocialModule.cs.patch]: On the note of "GetTemporaryFolderPath", the name is inaccurate - that folder is NOT temporary.
It is the folder where workshop files are stored prior to upload and persists across pushed updates for delta comparison by Steam.

I do notice now that the fact that folder is localized to a computer may create issues with the preview system in that if you publish on a new computer it will not have the history files of the mod. Something to look in to.

File: patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs (hunk 9 above)
Comment (Solxanich) [reply patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs]: the 'filename' string is still the full file name, ie includes parent folders. this would work with both subfolders or if we wanted to rename the tmods. 
Let's stick with subfolders in case other files crop up in some mods.

File: .github/workflows/build.yml [hunk 12]
```diff
@@ -113,6 +157,18 @@ jobs:
           gpg --quiet --batch --yes --symmetric --cipher-algo AES256 --passphrase="$SETUP_ASSETS_PASSPHRASE" tModLoader.zip
           echo ::endgroup::
           
+          echo ::group::Giving Purpose         
+          if [ $BranchName = '1.4-stable' ] ; then
```
Comment (Solxanich) [inline .github/workflows/build.yml]: Where is branch name defined?
This is probably the cause of build type bug I have

Comment (Chicken-Bones) [reply .github/workflows/build.yml]: `BranchName` variable inside msbuild is defined in `src/WorkspaceInfo.targets` which is populated when you launch the setup tool

File: patches/tModLoader/Terraria/ModLoader/Core/BuildProperties.cs [hunk 13]
```diff
@@ -230,7 +230,7 @@ private static IEnumerable<string> ReadList(string value)
 						writer.Write("side");
 						writer.Write((byte)side);
 					}
-					if (!BuildInfo.IsRelease) {
+					if (!BuildInfo.IsStable) {
 						writer.Write("beta");
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/Core/BuildProperties.cs]: Lets remove this property completely.

File: patches/tModLoader/Terraria/ModLoader/ModLoader.cs [hunk 14]
```diff
@@ -36,7 +36,7 @@ public static class ModLoader
 		public static bool AlphaWelcomed;
 		public static bool ShowFirstLaunchWelcomeMessage;
 
-		public static string versionedName => (ModCompile.DeveloperMode || !BuildInfo.IsRelease) ? BuildInfo.versionedNameDevFriendly : BuildInfo.versionedName;
+		public static string versionedName => (ModCompile.DeveloperMode || !BuildInfo.IsStable && !BuildInfo.IsPreview) ? BuildInfo.versionedNameDevFriendly : BuildInfo.versionedName;
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/ModLoader.cs]: `BuildInfo.Purpose == Dev`

File: patches/tModLoader/Terraria/ModLoader/ModLoader.cs [hunk 15]
```diff
@@ -374,16 +374,11 @@ internal static void MigrateSettings()
 			if (LastLaunchedTModLoaderVersion < new Version(0, 11, 7, 5))
 				showMemoryEstimates = true;
 
-			if (BuildInfo.Purpose == BuildInfo.BuildPurpose.Alpha && LastLaunchedTModLoaderAlphaSha != BuildInfo.CommitSHA) {
+			if (BuildInfo.Purpose == BuildInfo.BuildPurpose.Dev && LastLaunchedTModLoaderAlphaSha != BuildInfo.CommitSHA || LastLaunchedTModLoaderVersion != BuildInfo.tMLVersion) {
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/ModLoader.cs]: Should we change `LastLaunchedTModLoaderAlphaSha` to remove the `Alpha` and use it everywhere? The other places these are checked are a bit inconsistent with this one.

File: patches/tModLoader/Terraria/ModLoader/UI/ModBrowser/UIModDownloadItem.cs [hunk 16]
```diff
@@ -81,7 +81,7 @@ internal class UIModDownloadItem : UIPanel
 			_moreInfoButton.OnClick += ViewModInfo;
 			Append(_moreInfoButton);
 
-			if (!ModLoader.versionedName.Contains(ModDownload.ModloaderVersion)) {
+			if ((BuildInfo.IsStable || BuildInfo.IsPreview) && !ModLoader.versionedName.Contains(ModDownload.ModloaderVersion)) {
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/UI/ModBrowser/UIModDownloadItem.cs]: `!BuildInfo.IsDev`
But also, this is gonna break..
I see below you have `// Use the stable version of the mod for publishing metadata, not the preview version!`

1. Old mod versions should be assumed forward compatible
2. If a mod is new and only published on preview... should there be an appropriate message?

File: patches/tModLoader/Terraria/Program.TML.cs [hunk 17]
```diff
@@ -36,5 +35,35 @@ public static partial class Program
 					RuntimeHelpers.RunClassConstructor(type.TypeHandle);
 			}
 		}
+
+		private static void PortOldSaveDirectories() {
+			// Port old file format users
+			var oldBetas = Path.Combine(SavePath, "ModLoader/Beta");
+
+			if (Directory.Exists(oldBetas)) {
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Program.TML.cs]: Code style, prefer to invert if to reduce nesting here.

File: patches/tModLoader/Terraria/Program.cs.patch [hunk 18]
```diff
@@ -100,15 +104,23 @@
...
++
++			SavePath = Path.Combine(SavePath, fileFolder);
 +
 +			if (File.Exists("savehere.txt"))
-+				SavePath = "ModLoader"; // Fallback for unresolveable antivirus/onedrive issues. Also makes the game portable I guess.
++				SavePath = fileFolder; // Fallback for unresolveable antivirus/onedrive issues. Also makes the game portable I guess.
 +
-+			//EnsureMinimumStackSizeOnThreads();
++			if (LaunchParameters.ContainsKey("-tmlsavedirectory"))
++				SavePath = LaunchParameters["-tmlsavedirectory"];
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Program.cs.patch]: Extract method. Move after logging init. Add a log message with current save path.

File: patches/tModLoader/Terraria/Social/Steam/WorkshopSocialModule.TML.cs [hunk 19]
```diff
@@ -33,6 +33,11 @@ public partial class WorkshopSocialModule
 				return false;
 			}
 
+			if (!BuildInfo.IsStable && !BuildInfo.IsPreview) {
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Social/Steam/WorkshopSocialModule.TML.cs]: `BuildInfo.IsDev`

File: patches/tModLoader/Terraria/Terraria.csproj.patch [hunk 20]
```diff
@@ -20,10 +20,10 @@
...
 +  <PropertyGroup>
-+    <tMLVersion>0.12</tMLVersion>
++    <tMLVersion Condition="'$(TmlVersion)' == ''">9999.0</tMLVersion>
 +    <BuildPurpose Condition="'$(BuildPurpose)' == ''">dev</BuildPurpose>
 +    <BranchName Condition="'$(BranchName)' == ''">unknown</BranchName>
-+  <BranchName>$(BranchName.Replace("-",""))</BranchName>
++    <BranchName>$(BranchName.Replace("-",""))</BranchName>
 +    <CommitSHA Condition="'$(CommitSHA)' == ''">unknown</CommitSHA>
 +    <BuildDate>$([System.DateTime]::UtcNow.ToBinary())</BuildDate>
 +    <SourceRevisionId>$(tMLVersion)-$(BranchName)-$(BuildPurpose)-$(CommitSHA)-$(BuildDate)</SourceRevisionId>
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Terraria.csproj.patch]: Can we change the separator to `|` so that we don't have to replace `-` in branch names?

File: patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs [hunk 21]
```diff
@@ -319,5 +325,77 @@ internal static class ModOrganizer
 				return new HashSet<string>();
 			}
 		}
+
+		private static readonly Regex PublishFolderMetadata = new Regex(@"([0-9][0-9][0-9][0-9]*[.][0-9]*)");
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs]: Regex should be more strict?

File: patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs [hunk 22]
```diff
@@ -319,5 +325,77 @@ internal static class ModOrganizer
...
+
+		internal static string GetActiveTmodInRepo(string repo) {
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs:331]: Can we refactor a couple of these around a method which returns `IEnumerable<(Version majorMinor, string path)>`

**Review by Chicken-Bones — CHANGES_REQUESTED** [review]
Looking really good!

Main thing I still want to do is allow a null tML version.

**Comment by Solxanich** [comment]
Comment from Discord: Cleanup the comments in the yaml files to actually be useful

File: patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs (hunk 21 above)
Comment (ExterminatorX99) [inline patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs]: Is the regex meant to match on the following? It currently does. 
`2022222.`
`202.`
`202.222222222222222`
`2022.2222222`
`2022000000.222222222`

File: patches/tModLoader/Terraria/ModLoader/ModLoader.cs (hunk 15 above)
Comment (Solxanich) [reply patches/tModLoader/Terraria/ModLoader/ModLoader.cs]: I'll have to defer to Jopo on this one for eliminating the LastLaunchedAlpha fully - the setup for how the "Show What's New" prompt is quite integrated with it, so I'd rather not break that right now

File: patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs [hunk 23]
```diff
@@ -326,7 +326,7 @@ internal static class ModOrganizer
 			}
 		}
 
-		private static readonly Regex PublishFolderMetadata = new Regex(@"([0-9][0-9][0-9][0-9]*[.][0-9]*)");
+		private static readonly Regex PublishFolderMetadata = new Regex(@"[/|\\]([0-9][0-9][0-9][0-9][.][0-9][0-9]*)[/|\\]");
```
Comment (ExterminatorX99) [inline patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs]: This regex matches on `|2022.22222222|`

I get the feeling the regex you want is
`[/|\\]([0-9]{4}[.][0-9]{1,2})[/|\\]`

Comment (Solxanich) [reply patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs]: Ok, will use this. Thanks for hammering that out

File: patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs [hunk 24]
```diff
@@ -319,5 +325,77 @@ internal static class ModOrganizer
 				return new HashSet<string>();
 			}
 		}
+
+		private static readonly Regex PublishFolderMetadata = new Regex(@"[/|\\]([0-9]{4}[.][0-9]{1,2})[/|\\]");
```
Comment (ExterminatorX99) [inline patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs:329]: Here's an even stricter regex that won't match on invalid dates like `|2022.23|`

`[/|\\]([0-9]{4}[.](0?[1-9]|1[0-2]))[/|\\]`
